Update `aws-c-common`

excitoon opened the pull request:
**Please merge with fast-forward without merge commit.**

danielsn committed:
Add additional CBMC checks by default (#628)

danielsn committed:
remove unnecessary uses of __CPROVER_allocated_memory (#627)

Commit:
Added small block allocator (#625)

* Added small block allocator

* Added clz/ctz impls for all compilers

* Added thread safety controls to sba allocator

* Implemented clz/ctz for common integer types, both signed/unsigned

singku-china committed:
Add utility to append null terminator to aws_byte_buf.

Co-authored-by: Andrew Tang <tangliang551@gmail.com>

Commit:
Windows compilation fixes (#631)

* Use _WIN32 consistently, no WIN32

* Use signed compare in fallback ctz implementations

* Added cross compile CI

* Don't build consumers for cross compile, they're not all guaranteed to work

* Use builder version v0.6.4

TingDaoK committed:
Caches (#634)

* Revert "Remove unused linked_hash_table datastructure (#490)"

This reverts commit 049b4812da213644523d8ccd1983fa444928ea05.

* FIFO/LIFO/LRU caches based on aws_linked_hash_table
* APIs for LRU cache is changed here


Co-authored-by: Dengke Tang <dengket@amazon.com>

TingDaoK committed:
Cache destroy (#636)

- clean-up -> destroy
Co-authored-by: Dengke Tang <dengket@amazon.com>

JonathanHenson committed:
IO handle ulimit stuff. (#637)

API for reading max fds and updating the soft limit.

Commit:
Moved use of GCC builtin intrinsics to their own math inline file (#639)

* Moved use of GCC builtin intrinsics to their own math inline file

* Disable fallbacks for ctz/clz when intrinsics are available

bretambrose committed:
Newline (#640)

JonathanHenson committed:
Added newline to process_test.c (#641)

* Added newline to process_test.c

graebm committed:
aws versions of isspace() isalpha() etc (#642)

Versions of isspace() isalpha() etc for use with ASCII/UTF-8.
C's built-in functions are affected by user's locale.

graebm committed:
clarify naming with examples (#643)

kankri committed:
Support STATIC_CRT setting for MSVC (#638)

Co-authored-by: Risto Kankkunen <risto.kankkunen@f-secure.com>
Co-authored-by: Jonathan M. Henson <henso@amazon.com>
Co-authored-by: Justin Boswell <boswej@amazon.com>

Commit:
Doc update for STATIC_CRT (#644)

JonathanHenson committed:
Manual cpu feature checks to avoid llvm linking problems due to __cpu_model (#645)

* removed built-in-cpu supports so llvm will quit complaining at link time about __cpu_model

Commit:
Added BOM detection/encoding detection via aws_text_encoding_detect (#647)

* Added BOM detection/encoding detection via aws_text_encoding_detect

* Added aws_text_is_utf8 and tests, as well as ascii resource test

bretambrose committed:
Bigint updates - constant time and api changes (#646)

*  Updates Bigint interface to be more consistent
*  Changes some Bigint constructors to constant-time
*  Changes Bigint comparison to constant-time


Co-authored-by: Justin Boswell <boswej@amazon.com>

singku-china committed:
Update license to remove Copyright year. (#648)

Co-authored-by: Andrew Tang <tangliang551@gmail.com>

karkhaz committed:
Move CBMC proofs to tests/cbmc

This commit restructures the CBMC proof directories:

* The overall proof directory `.cbmc-batch` -> `tests/cbmc`
* The directory containing the proofs `jobs` -> `proofs`
* The directory containing proof source file `source` -> `sources`

This is in preparation for commits that will use a common unified
Makefile for CBMC proofs.

karkhaz committed:
Add aws-templates-for-cbmc-proofs submodule

This commit submodules in the CBMC Makefile Template repository, hosted
at

    https://github.com/awslabs/aws-templates-for-cbmc-proofs

This repository contains templates for Makefiles and associated files
for doing CBMC proofs. This commit is in preparation for migrating the
aws-c-common proofs to use this unified Makefile, which will simplify
proof setup and maintenance.

bretambrose committed:
Remove bigint (#650)

bretambrose committed:
Add aws_byte_buf_append_dynamic_secure (#649)

* Add aws_byte_buf_append_dynamic_secure
* Added byte appends

karkhaz committed:
Set C-Common specific defines for CBMC Makefile

This commit adds definitions, include paths, and other Make variables
that are specific to C-Common to the CBMC proofs subdirectory. These
variables are necessary to ensure that proofs build correctly.

karkhaz committed:
Make CBMC proof Makefiles conform to template

This commit changes the variable names in each CBMC proof's Makefile so
that they conform to the naming scheme in the template Makefile.common.
The changes are as follows:

* `ENTRY` -> `HARNESS_ENTRY`.
* The name of the proof file is `HARNESS_FILE`.
* `HARNESS_FILE` is added to the `PROOF_SOURCES` list.
* All `DEPENDENCIES` that are under the cbmc/sources directory are now
  listed under the `PROOF_SOURCES` variable, and their directory root is
  set to `PROOF_SOURCE`.
* All `ABSTRACTIONS` that are under the cbmc/stubs directory are now
  listed under the `PROOF_SOURCES` variable, and their directory root is
  set to `PROOF_STUB`.
* All `DEPENDENCIES` that are under the c-common source root are now
  listed under the `PROJECT_SOURCES` variable, and their directory root is
  set to `SRCDIR`.
* `REMOVE_FUNCTION_BODY` now contains only function names, without the
  `--remove-function-body` flag preceding each function

karkhaz committed:
Move tests/cbmc to verification/cbmc

karkhaz committed:
Fix inclusion paths for CBMC proofs

karkhaz committed:
Add empty source file for .inl-using proofs

Some proofs in C-Common don't link against any source files at all,
because the code that they are proving is written in .inl files that are
included by stubs. This commit adds an empty C file as the
PROJECT_SOURCES for those proofs so that the linking process proceeds
correctly.

feliperodri committed:
Adds correct nondet_uint8_t() for aws_array_list_swap stub

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

AGSaidi committed:
Add arm cpuid detection (#653)

graebm committed:
aws_byte_buf_write_to_capacity() (#656)

karkhaz committed:
Bump Makefile template for nondet-static bugfix

This commit bumps the submodule pointer of the template Makefile in
order to get the bug fix merged in
https://github.com/awslabs/aws-templates-for-cbmc-proofs/pull/6

karkhaz committed:
Add declaration for uninterpreted function

This commit fixes two proofs that had assumed that the return type of an
uninterpreted function was `int`. These proofs break when using the
broader range of CBMC checks that are introduced by the common Makefile.

karkhaz committed:
Bump Makefile template submodule pointer

This is so that we can use the .gitignore files that were added in
https://github.com/awslabs/aws-templates-for-cbmc-proofs/pull/5

karkhaz committed:
Use correctly mangled loop names in CBMC proofs

This commit fixes the name of a loop in several linked_list CBMC proofs.
The loop name had previously been working correctly due to a bug in
CBMC, which has now been addressed. When using up-to-date versions of
CBMC, the changes in this PR will be required.

karkhaz committed:
Add .gitignore from template Makefile

This commit adds the standard .gitignore from the CBMC template
Makefile, ensuring that all CBMC output directories are ignored.

graebm committed:
aws_linked_list_move_all_back() (#664)

O(1) helper functions to move all nodes from one list to another.

This is something we do pretty commonly while holding a lock. It's much less code using these helper functions.

Commit:
Win7 support (#659)

* Disable secure CRT warnings in test code for MSVC

* Always target Win7 API with MSVC

* Removed ./ from test file paths for encoding tests

* Fix working dir on windows tests

Commit:
Added AWS_SUPPORT_WIN7, Win7 is no longer supported by default (#669)

* Added AWS_SUPPORT_WIN7, Win7 is no longer supported by default

* Added AWS_SUPPORT_WIN7 macro for feature detection

bretambrose committed:
Add new guideline with respect to checking errors and logging (#671)

bretambrose committed:
Don't include avx2 and stuff on ollllllld compilers (#655)

rccarper committed:
Moving XML parser from aws-c-auth into aws-c-common (#674)

* Moving XML parser from auth into common

* Making xml_parser and xml_attribute opaque and adding to interface

* Adding usage of aws_xml_get_num_attributes to test

graebm committed:
aws_byte_cursor_read_hex_u8() and other string helpers (#657)

**new**  `aws_byte_cursor_read_hex_u8()`: Reads 2 hex characters from ASCII/UTF-8 text to produce an 8-bit number.

**new** `aws_lookup_table_hex_to_num_get()`:  Returns lookup table to go from ASCII/UTF-8 hex character to a number (0-15).

**new** `aws_string_new_from_cursor()` and `aws_string_new_from_buf()`: obviously

**fixed** `aws_byte_cursor_next_split()`: fixed to handle source with 0 len and NULL ptr

graebm committed:
aws_byte_buf_init_cache_and_update_cursors() (#676)

Helper function to init an `aws_byte_buf` that backs up a bunch of cursors.

feliperodri committed:
Upgrades Makefile templates version for CBMC proofs (#679)

* Upgrades Makefile templates version
* Disables CBMC proofs with CI failures
* Fixes loop-unwinding failures

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

SaswatPadhi committed:
Update CBMC proof harnesses (#678)

This change reenables and fixes the 12 proofs that were disabled in #679.

Commit:
Added AWS_OS_* macros to detect OS (#681)

* Added AWS_OS_* macros to detect OS

* Added platform.h to common.h

* Fixed narrowing warning for backtrace APIs

Commit:
Fixed iOS check (#682)

danielsn committed:
use default cbmc-batch (#684)

bretambrose committed:
String to log level (#687)

* String to log level

bretambrose committed:
Ref counting helpers (#675)

* Ref counting APIs
* Global thread creator waiter

feliperodri committed:
Improves aws_byte_buf validity functions (#686)

* Improves aws_byte_buf validity functions
* Avoid invalid aws_byte_buf in unit tests
* Makes ensure_byte_buf_has_allocated_buffer_member nondet
* Updates s_byte_cursor_limit_tests_fn unit test

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

feliperodri committed:
Updates post-conditions in aws_mul_size_checked proof (#689)

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

AGSaidi committed:
Add arm64 inline assembly for saturating ops (#670)

Co-authored-by: Jonathan M. Henson <henso@amazon.com>

AGSaidi committed:
Use -moutline-atomics if available on Arm systems (#672)

The -moutline-atomics compiler flas generates code for both older Arm load/store
exclusive atomics and also Arm's Large System Extensions (LSE) which can scale
an order of magnitude better on large core count systems and are fair unlike ldrex/strex.

Co-authored-by: Jonathan M. Henson <henso@amazon.com>

bretambrose committed:
Why didnt this fail to build? (#690)

Commit:
Removed 0unstable SOVERSION (#691)

Commit:
Imported PR from #695 to fix aarch64 builds (#696)

feliperodri committed:
Improves predicates for aws_array_list proofs (#694)

* Adds AWS_IMPLIES and AWS_IFF macros
* Improves the aws_array_list validity function
* Improves predicates for CBMC proofs
* Fixes format-check.sh script to scan verification dir

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

graebm committed:
header checker checks for include-guards (#699)

graebm committed:
Print CMake version when configuring. (#701)

Useful for diagnosing issues on various machines.

rccarper committed:
Re-adding SOVERSION to CMakeLists.txt (#702)

* Re-adding SOVERSION to CMakeLists.txt to fix SONAME.

danielsn committed:
only memcpy if we have bytes (#704)

graebm committed:
shorter submodule paths (#705)

Shorten file path, as well as "logical path" in .gitmodules file. This is to aid with pain on windows when paths exceed 260 characters. We do a lot of recursive submodules, and these things add up.

Co-authored-by: Kareem Khazem <karkhaz@amazon.com>

feliperodri committed:
Modify CBMC proofs to make assumptions about malloc explicit (#708)

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

tegansb committed:
Updates aws_hash_table stubs for CBMC proofs (#716)

tegansb committed:
Updates to CBMC stubs and ensure_allocated_hash_table needed for malloc-may-fail flag and ESDK proofs  (#719)

* Added non-NULL assumptions to hash_iterator_string_string_generator and hash_find_string_string_generator

* Changes in stubs and ensure functions to handle malloc-may-fail flag

tegansb committed:
minor update to ensure_byte_buf_has_allocated_buffer_member (#721)

DavidOgunsAWS committed:
Adding aws-c-iot to documented reserved error code range. (#722)

tegansb committed:
Updates CBMC constructors for aws_bye_cursor and aws_hash_table (#724)

* minor update to ensure_byte_buf_has_allocated_buffer_member

* Updates to aws_byte_cursor and make_hash_table_no_backing_store

* Address PR comments

* can_fail_malloc updated to bounded_malloc

bretambrose committed:
Simple build os detection function (#725)

* Simple build os detection function; potentially improve granularity, flavors, etc... later

Commit:
Update to builder v0.7.0 (#726)

* Update to builder v0.7.0

* Replaced master with main

* Use generator for target output dir for test resources

* Fixed clang-tidy

* clang-format

JonathanHenson committed:
CPU affinity and memory policy support (kinda) (#727)

thread cpu affinity with some membind support for numa

feliperodri committed:
Remove unnecessary scrips from CBMC dir (#730)

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

bretambrose committed:
aws_byte_buf_append_n (#731)

* aws_byte_buf_append_n

feliperodri committed:
Uses Litani build system in CBMC proofs (#732)

* Upgrade CBMC templates
* Adds Litani submodule
* Updates .gitignore to include .ninja_logs
* Adds new Litani script
* Updates all Makefiles to run with Litani
* Replaces cbmc-proof.txt with cbmc-batch.yaml
* Gives run-cbmc-proofs.py the correct permission

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

Commit:
Disabled MSVC 2015's non-constant aggregate initializer warning (#738)

Commit:
Do not use addr2line by default, backtrace_symbols will do the job (#739)

* Do not use addr2line by default, backtrace_symbols will do the job

* Fixed memtrace/stack decoding test so it works more robustly

bretambrose committed:
Remove redundant, unused API (#740)

graebm committed:
Fix thread.c for FreeBSD (#741)

Commit:
Turned on -fno-omit-frame-pointer always, pointless optimization and greatly improves profiling (#743)

Commit:
Use aligned allocs for > PAGE_SIZE allocations, avoid unaligned memcpy/memset on large buffers (#742)

* Use aligned allocs for > PAGE_SIZE allocations to avoid unaligned memcpy/memset

* Fixed memory allocation of symbol blobs on windows

* Fixed symbol free on windows in memtracer

JonathanHenson committed:
Task scheduler with a background thread. (#744)

* Task scheduler with a background thread.

bretambrose committed:
Task asan fix (#745)

* Fix asan issue with tasks

Commit:
Fixed default alignment for small allocs (#746)

JonathanHenson committed:
Unixy numa discovery, still need windows. (#748)

numa discovery for cpu pinning for systems using libnuma (dynamically bound).

bretambrose committed:
TryLock was broken on Windows (#750)

* TryLock was broken on Windows
* Disable hanging test for now

Commit:
Disabled non-standard extension warning that is actually a lie on MSVC (#751)

* Disabled non-standard extension warning that is actually a lie on MSVC

JonathanHenson committed:
Fixed infinite loop in task cancelation. (#752)

* Fixed infinite loop in task cancelation.
* Make timeout something more reasonable to accomodate missed signals in the condition variable.

JonathanHenson committed:
set_mempolicy in libnuma is defined as a WEAK symbol. dlload() does not care or honor this value, libpthread links against libnuma.so and defines the symbol, madness ensues. (#757)

rccarper committed:
Fixing Clang Tidy (#759)

* Changing clang tidy github script to check 20.04 explicitly.
* Applying fixes from ThreadJoin branch for clang-tidy on newer version of Ubuntu.
* Adding check to clang-tidy also found in ThreadJoin branch.

TingDaoK committed:
OOS error (#761)

mapping ENOSPC to AWS_ERROR_NO_SPACE for out of space error

bretambrose committed:
Fix bug in date time accidentally introduced while trying to fix clang tidy (#760)

Co-authored-by: Michael Graeb <graebm@amazon.com>

bretambrose committed:
Thread join (#756)

* Initial posix managed thread implementation
* Platform-shared thread management

graebm committed:
fix shared libs (#762)

`aws_thread_set_managed_join_timeout_ns()` and `aws_thread_join_all_managed()` are declared in `thread.h` and implemented in `thread_shared.c`. But since `thread_shared.c` did not include `thread.h`, it was not seen as the implementation by the linker.

Commit:
Fixed documentation typo in aws_use_package docs (#765)

JonathanHenson committed:
Pthread np (#766)

Fix musl build:

Some platforms, such as musl, do not have support for
the nonportable pthread_attr_setaffinity_np function.
Some of those without may be able to fall back to
pthread_setaffinity_np.

We can directly check for support using CMake's CheckSymbolExists,
and enable/disable this functionality accordingly.

This is intended to be extensible to other methods,
such as OS X 10.5 mach thread policies.

Authored-by: Ryan Burns <rtburns@protonmail.com>

bretambrose committed:
Add aws-c-s3 to log subject and error code table (#769)

bretambrose committed:
Add a success/fail return value to aws_thread_managed_join_all (#768)

graebm committed:
Add shared-libs CI job (#767)

Also fix windows-vc14 (x86) job so it actually uses vc14, and actually builds x86

bretambrose committed:
Tag internal function as static (#770)

bretambrose committed:
Clang updates (#773)

* Clang updates - add 10 and 11, fix <= 8

bretambrose committed:
Mingw2 (#772)

* Windows: Check _MSC_VER for MSVC specific pragma warnings directives.
* include/common: Check _WIN32 to prevent htonl and ntohl functions.
* tests: Check if alloca is already defined.

Co-authored-by: Biswapriyo Nath <nathbappai@gmail.com>
Co-authored-by: Joseph Klix <jkl@amazon.com>

xhochy committed:
Don't hardcode the CoreFoundation path (#733)

* Don't hardcode the CoreFoundation path

This would otherwise hardcode the Xcode version and we would have to recompile after each Xcode update.

Fixes #510

* Address review comments

Co-authored-by: Jonathan M. Henson <henso@amazon.com>
Co-authored-by: Bret Ambrose <bretambrose@gmail.com>

bretambrose committed:
Add asan workflow (#775)

feliperodri committed:
Updates readable and writable macros (#778)

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

rccarper committed:
Fixing Thread Affinity CMake (#779)

feliperodri committed:
Prevents undefined behavior in allocators (#780)

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

bretambrose committed:
Disable warning on test (#781)

* Disable stringop-overflow warning when building tests since sometimes cmake generates code that violates that warning and then downstream consumers get unhappy

bretambrose committed:
Set log level (#782)

* Add support in the default logger for changing the current log level dynamically.

Co-authored-by: Michael Graeb <graebm@amazon.com>

bretambrose committed:
Move sanitizer builds to builder jobs (#784)

bretambrose committed:
Testing arg passing to builder (#786)

* Fix sanitizer argument passing

Commit:
Applied changes to support non-desktop windows API families (#788)

* Applied changes to support non-desktop windows API families

* Moved windows probe to feature detection at configure time

graebm committed:
Prevent log-subject ordering errors (#789)

bretambrose committed:
Avoid overflow on Windows high resolution timer queries  (#791)

* Avoid overflow on Windows high resolution timer queries once system has been up long enough

bretambrose committed:
Clock2 (#793)

* Fix Windows high resolution clock issue by updating and using the time conversion API

feliperodri committed:
Removes all proof allocators from CBMC proofs (#783)

* Removes all proof allocators from CBMC proofs
* Includes allocator.c in CBMC proofs

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

feliperodri committed:
Upgrades CBMC templates for proof harnesses (#795)

* Upgrades CBMC templates for proof harnesses
* Upgrade to Litani 1.6.0 for CBMC proofs

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

TingDaoK committed:
Aws fopen (#794)

introduce aws_fopen to support the non-ascii filepath on windows

Commit:
Diagnostics fixes/improvements (#796)

* Disable AddressSanitizer on hashlittle2, it overreads, but not in a dangerous way

* Added log level to aws_backtrace_log, and fixed stacktrace decoding test to be permissive on weird platforms

* Fixed aws_backtrace_log versions for execinfo/no execinfo

sdavtaker committed:
Marking aws-c-common/source/posix/system_info.c as a non parameters function (#797)

DavidOgunsAWS committed:
Adding CMake option to disable warnings as errors (#798)

* Adding cmake option to disable Werror. On by default.

DavidOgunsAWS committed:
Renaming flag to avoid clashes. (#799)

DavidOgunsAWS committed:
Compiler warnings are not errors by default normal build success (#800)

* Compiler warnings are not errors by default normal build success

Commit:
Added builds for arm32v5 and arm32v7 (#787)

* Added builds for arm32v5 and arm32v7

* Added qemu install

* test with arm branch of builder

* Updated to latest builder release

Co-authored-by: Bret Ambrose <bretambrose@gmail.com>

TingDaoK committed:
Avoid null-pointer arithmetic (#802)

* Avoid arithmetic operations over NULL pointers
* Disables pointer-overflow check as reasoning about deallocated object
* Instead of AWS_MEM_IS_READABLE, check for nullness
* Avoid unreasonable allocation
* Avoid null-pointer arithmetic in aws_ring_buffer
* Upgrades CBMC templates
* Updates .gitignore for CBMC proofs
* Makes clang-format happy again
* comments addressed
* further comments
* typo and just check allocation_end should be enough
* two of them are cursor...
* finally get the report. and everything should work now...

Co-authored-by: Felipe R. Monteiro <felisous@amazon.com>

graebm committed:
Improved Posix Large File Support (LFS) (#808)

`aws_off_t` is now always 64 bits. It's also deprecated, `int64_t` should be used instead instead.

### ISSUE:
Bizarre behavior on 32bit Raspberry Pi OS when aws_input_stream_seek() was called.

### CAUSE:
Conditional logic in header files is bad.

The size of `off_t` on Posix is conditional, based on various defines at when the headers are included.

`aws_off_t` aliased `off_t`

It's possible for .c files in different libraries to disagree on the size of `off_t`, which means they disagree on the size of `aws_off_t`, which leads to stack corruption when functions taking `aws_off_t` are called.

### CHANGES:
1) Deprecate aws_off_t and redefine it to always be 64 bits. We'll deal with 32 vs 64 bit off_t in .c files we control, not in headers. I will fix up downstream libraries in followup PRs.
2) AwsCFlags now attempts to enable LFS, even on 32 bit system, by setting `-D_FILE_OFFSET_BITS=64` if necessary. This secretly fixes bugs we didn't even know we had: like `fopen()` now  uses `fopen64()` under the hood.

Commit:
Added aws_promise and tests (#807)

* Added aws_promise and tests

* Fixed clang-format source dir

graebm committed:
Fix Android Large File Support checks (#811)

32bit Android can have 64bit off_t, but no fseeko().

The check_c_source_compiles() sample program was supposed to check for this, but the missing fseeko() was only treated as a compiler warning and didn't fail the check. Therefore, explicitly check for fseeko as a separate check. (I couldn't figure out how to get check_c_source_compiles() to do warnings-as-errors)

DavidOgunsAWS committed:
Adding cmake include for check_symbol_exists() now used in aws-c-io (#812)

feliperodri committed:
Fix vacuity check in aws_priority_queue_s_swap proof (#815)

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

Commit:
Only apply -fno-omit-frame-pointer when debug info is present (#818)

Commit:
Added a function to split debug info out of shared libs and executables (#819)

kankri committed:
Use formatting specifiers correctly (#820)

Co-authored-by: Risto Kankkunen <risto.kankkunen@f-secure.com>

bretambrose committed:
Part 1 of aws_thread_clean_up fix and contract clarification (#814)

* Part 1 of aws_thread_clean_up fix and contract clarification
* Close the original windows handles too

graebm committed:
Improved fix for string-overflow warning. (#821)

The previous attempt to fix this warning required both AwsFeatureTests.cmake and AwsTestHarness.cmake to be included in every CMakeLists.txt to work properly. With this change, the fix works without including both files.

bretambrose committed:
Clang mingw (#822)

* cmake: Append -moutline-atomics for AArch64 only. -moutline-atomics only available for AArch64 or ARM64 only. See https://gcc.gnu.org/onlinedocs/gcc/AArch64-Function-Attributes.html
* cmake: Check -fPIC compiler flag before appending.  clang for x86_64-w64-windows-gnu target does not support -fPIC flag.
* source: Include errno.h for EINVAL.

Co-authored-by: Biswapriyo Nath <nathbappai@gmail.com>

yourslab committed:
Update datetime to support millisecond precision (#824)

* Update datetime to support millisecond precision

* Add unit tests for millis and doubles

bretambrose committed:
Add dockerfile for the dotnet crt CD build (#823)

yourslab committed:
Have aws_date_time_as_epoch_secs return milliseconds as the decimal portion (#825)

Function was not updated to return the milliseconds as the decimal portion of the double

karkhaz committed:
Upgrade CBMC proof tools: starter kit and Litani 1.10.0 (#826)

* Upgrade proof tool submodules

This commit advances Litani to release 1.10.0, and the starter kit to
the tip-of-tree. This brings the following improvements:

- Profiling
    - Litani measures the memory usage of the CBMC safety checking and
      coverage checking jobs
    - The dashboard includes box-and-whisker diagrams for memory use per
      proof
    - The dashboard includes a graph of how many parallel jobs are
      running over the whole run, making it easy to choose a CI machine
      with enough parallelism
    - It is now possible to designate particular proofs as "EXPENSIVE";
      Litani runs expensive proofs serially, ensuring that they do not
      over-consume resources like RAM.

- UI improvements
    - Each pipeline page includes a table of contents
    - Each pipeline page includes a dependency graph of the pipeline
    - Each job on the pipeline page has a hyperlink to that job
    - The terminal output is now less noisy

* Change cbmc-batch.yaml to cbmc-proof.txt

This makes the proof layout consistent with the starter kit, which will
allow us to use a generic run script in a future commit. Putting this
in commit by itself because the diff is huge and not worth reading (just
moving some files and changing two lines in the runscript).

* Symlink run-cbmc-proofs.py to starter kit

The run script is now a symbolic link into the starter kit submodule,
meaning that it will be updated whenever the starter kit is. This is
done iso that AWS C Common doesn't carry custom modifications to the run script
unless necessary; previous commits have made the AWS C Common proofs consistent
with the generic starter kit conventions.

kkang-wr committed:
priority_queue.c: fix compile error with -Og (#828)

It fails to compile with gcc option '-Og':

| aws-c-common/0.5.11-r0/git/source/priority_queue.c:116:13:
   error: 'parent_item' may be used uninitialized in this function [-Werror=maybe-uninitialized]
|   116 |         if (queue->pred(parent_item, child_item) > 0) {
|       |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
| cc1: all warnings being treated as errors

Initialize variables with NULL to fix the issue.

Signed-off-by: Kai Kang <kai.kang@windriver.com>

feliperodri committed:
Fixes specifications for aws_hash proofs (#827)

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

SaswatPadhi committed:
Remove `memcmp` stubs (#831)

TingDaoK committed:
callback error out aws_hash_table_foreach (#829)

Commit:
Added leak assert to aws_sba_allocator's destroy path (#832)

Commit:
Added metrics to SBA and tests for metrics (#833)

bretambrose committed:
Migrate to ecr and update/consolidate CI (#837)

* Migrate to ecr and update/consolidate CI

feliperodri committed:
Upgrade CBMC sub-modules (#838)

* Upgrade to Litani 1.11.1
* Upgrade CBMC standard files

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

Commit:
Added aws_bus (#836)

ilevyor committed:
allow use of byte_buf_reserve on buffers with initial capacity 0 (#843)

* allow use of byte_buf_reserve on buffers with initial capacity 0

graebm committed:
-DUSE_CPU_EXTENSIONS=OFF disables CPU-specific function implementations (#842)

**Issue:**
We have customers encountering mysterious crashes in some of these functions. Give them the ability to disable them while we attempt to diagnose the issue.

**Description of changes:**
New build flag controls whether to use optimized functions that make use of CPU extensions (ex: AVX, SSE).

graebm committed:
Fix FAIL() macro (#845)

In MSVC, the FAIL() macro didn't compile when there were no additional args after the format string

TingDaoK committed:
Array list push front (#846)

Commit:
Updated README table of library ids (#848)

Commit:
Imported file utils from IO (#841)

* Imported file utils from IO

* Fixed non-LFS seek

* Fixed aws_file_get_length on windows

* Windows needs Shlwapi

Commit:
Removed ARN parser from common, it's in sdkutils now (#851)

* Removed ARN parser from common, it's in sdkutils now

* Removed tests for arn parser

TingDaoK committed:
Adds a proof for aws_array_list_push_front (#847)

add proof and bug fix

Co-authored-by: Felipe R. Monteiro <felisous@amazon.com>

JonathanHenson committed:
Directory traversal support (#830)

mem_acquire no longer allows returning NULL, Windows unicode support in aws_string.h, file system and directory implementations for windows and posix.

Commit:
Only attempt to use sys/auxv.h if it exists (#852)

JonathanHenson committed:
windows string fixes. (#853)

Commit:
Fixed encoding_avx2.c being compiled when AVX is not detected (#854)

JonathanHenson committed:
Make sure check_c_compiles fails on compiler warnings. (#859)

graebm committed:
Fix crash on machines that claim to support AVX2 but not AVX (#858)

bretambrose committed:
Disable CPU extensions on mingw64 builds on Windows (#862)

* Disable CPU extensions on Mingw-based builds.  Compiler bug in gcc leads to misaligned avx2 instructions leading to crashes

Commit:
Added support for detecting watchOS and tvOS (#863)

feliperodri committed:
Removes unused stubs from CBMC proofs (#867)

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

SaswatPadhi committed:
Make byte_buf proof allocator consistent with code (#869)

bretambrose committed:
Remove aws_ring_buffer_allocator_init declaration from public header (#870)


Co-authored-by: Hassan Sahibzada <hsahibza@amazon.com>

feliperodri committed:
Upgrade CBMC submodules (#871)

* Upgrade Litani version
* Upgrade CBMC templates submodule

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

feliperodri committed:
Bump Litani version to 1.16.0 (#872)

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

TingDaoK committed:
Support for Windows server 2008 (#868)

Issue #, if available:

-    Functions we used not existed in windows server 2008

Description of changes:

-    Try to load the function from kernel during runtime.
-    If not exist, either not improve it or just return error
-    Raise error when the function is inevitable

JonathanHenson committed:
added subcommand dispatch for cli argument parser. (#877)

JonathanHenson committed:
Improved cli parsing to handle positional arguments better so they ca… (#878)

* Improved cli parsing to handle positional arguments better so they can be handled as encountered, added state reset api.

bretambrose committed:
Release lock on no-alloc fwrite error path (#880)

cmello committed:
Add aws_byte_cursor_utf8_parse_u64 (#881)

This commit is the first step of the following refactor:

* Move the aws_strutil_read_unsigned_num function from aws-c-http to aws-c-common (renamed to aws_byte_cursor_utf8_parse_u64)
* Delete the function definition in aws-c-http and the corresponding copy aws_s3_strutil_read_unsigned_num in aws-c-s3
* fix callers of the function in aws-c-http and aws-c-s3 so they use the version from aws-c-common

graebm committed:
Add AWS_ERROR_PLATFORM_NOT_SUPPORTED error. (#882)

In my opinion, this is more clear than using AWS_ERROR_UNIMPLEMENTED for features that can't work on a given platform.

TingDaoK committed:
CI use tag instead of latest incase latest change and breaks us (#886)

TingDaoK committed:
refcount check the old value before acquire as well (#890)

bretambrose committed:
Update builder to latest (#894)

* Update builder to latest

SergeyRyabinin committed:
Append pthread to platform libs if cmake found Thread library does not provide required symbol (#897)

Co-authored-by: Sergey Ryabinin <srryabin@amazon.com>

graebm committed:
Fix "unused variable" compiler warning (#899)

TwistedTwigleg committed:
Add cJSON module (#889)

Moving CJSON into a common module so it can be effectively reused across
the C CRT repositories without having to embed multiple copies of CJSON.

Commit log:
* Added CJSON module implementation
* Adjusted JSON module to fit code style, added code documentation
* Fixed formatting issues, added clang-format files to skip formatting cJSON
* Added files to ignore clang-tidy in externals directory
* Further adjustments to work around clang-tidy on external, adjustment to fix compiling on Windows
* More formatting fixes
* JSON module header clang-tidy fixes
* Initial design of AWS JSON module
* Redesign with new interface that does not expose cJSON
* Fixed the JSON module crashing.
* Added additional JSON testing
* Fixed formatting issues, used struct instead of void*
* Additional clang-format fixes
* Added documentation, further clang-format fixes
* Fixes for Windows CI
* Moved JSON code into source to attempt to fix Windows CI
* Modified to fit C repository style and to use C repo error style
* Fixed JSON tests
* Added AWS_COMMON_API to API functions
* Added missing AWS_COMMON_API
* Removed AWS_COMMON_API from struct in JSON module
* Clang-tidy fixes for CJSON
* Merged latest version of aws-c-common and testing including cJSON C file instead of header.
* Revert include change, fix ifndef
* Code review changes
* Further code review changes
* Upgraded cJSON to latest version
* Larger code review changes
* Adjusted cJSON version from main to latest release, adjusted CMAKE file to not include cJSON headers
* Fix warnings due to const casting
* Fixed issues with tests, minor API adjustment as needed to get everything working
* fixed compiling issue on CI, ran clang-format
* Another clang format fix and fixing lgtm issue
* Got the print functions no longer leaking memory
* Removed warnings of unused allocator parameter in JSON functions
* Code review changes to JSON module
* Additional code review changes and clang format pass
* Fixed issues where Windows was not building due to tests failing, hopefully fixed clang-tidy issues on cJSON files
* Fixed header and source for JSON module having different arguments causing vc14 to fail
* Further fixes for clang-tidy and vc14
* Adjusted code based on review
* Added additional checks before cleaning byte buffer
* Fixed lgtm bot issue and hopefully issue causing CI to fail
* Code review update: Naming adjusted and functionality slightly tweaked
* Adjusted JSON string functions to append to byte buffer. Naming adjustment
* Use secure byte buffer appends, fix test condition check order
* Fix JSON tests failing on Linux

graebm committed:
Fix it so CI runs on branches with '/' in name (#900)

TwistedTwigleg committed:
Added metric reporting functions (#898)

Adds metric reporting functions that are platform specific that can be used in the Device Defender custom metrics

Commit change log:
* Added metric reporting functions that are platform specific that can be used in the Device Defender custom metrics
* Fixed formatting issues and unused variable issues on Mac
* Additional Mac fixes
* Moved metrics reporting to system_info
* Clang format fixes
* Changed how CPU is reported so it is clearer, detects errors faster, and handles more situations
* Clang format and clang tidy fixes
* Fixed LGTM error and unused variable issue on Mac
* Added use of aws_*_u64_checked and removed debug print statement
* Clang format fix (I need to setup Clang-format on linux...)
* Process count and memory usage now reports using int64_t
* WIP initial CPU usage sampler
* Adjusted CPU usage sampler to fit correct design
* Clang format fixes
* Further clang-format fixes - missed a couple minor issues
* Code review changes - first half
* Code review part 2 - refactored cpu_usage_sampler into OS specific folders. Added stub/default implementation for Windows and MacOS
* Clang format fix
* Further clang-format fixes
* Code review changes: Moved generic API functions for reuse, simplified code, adjusted comments
* Fixed clang format issues, fixed Windows compiling and clang-tidy
* More clang-tidy fixes
* Adjusted after code review: Added private header file for CPU sampler, changed linux to cache result on constructor, unsupported platforms return null
* Fixed unused argument issue on Windows
* Fix header guard using incorrect format

graebm committed:
aws_byte_cursor_starts_with(input, prefix) (#905)

TwistedTwigleg committed:
Add check for if_link.h to support platforms without it, like RHEL5 (#907)

TwistedTwigleg committed:
Remove DeviceDefender custom metric reporting functions (#901)

Removes DeviceDefender metric gathering functions from aws-c-common.

Commit log:
* Adjusted to better fit code style, added Windows support for process and memory statistics, general code improvements
* Clang format fixes
* Clang-tidy fix
* Removed unnecessary clamp condition, added MacOS support for getting memory used and process count
* Fixed clang-format issue
* MacOS metric fixes, clang-format fixes
* Hopefully final clang-format fix
* Code review changes and cleanup
* Moved cpu_usage_sampler to system_info, renamed where needed to system_cpu_sampler, fixed clang format
* Further clang format fixes
* Header import fixes, clang-tidy fixes
* Further fixes - header fixes and missing imports
* Remove CPU, process, and memory usage metrics reporting functions
* Removed header includes that are no longer needed

graebm committed:
improved checks for `-moutline-atomics` (#909)

**Issue #:** https://github.com/awslabs/aws-c-common/issues/902
build failure for iOS that didn't occur until an application was finally linked together.

**Description of changes:**
Test by compiling a program that actually invokes atomic operations, instead of simply calling `check_c_compiler_flag()`. 

The errors didn't surface in the empty program CMake uses in `check_c_compiler_flag()` checks.

TingDaoK committed:
fix compiler warning (#910)

TwistedTwigleg committed:
Adjust logging function to fix segfault if filename and file are not set (#911)

* Adjusted aws_logger_init_noalloc to no longer assume a file name has been set. If no file or filename is set, use stderr and log a warning
* Remove the log and note the use of stderr as the default in the header file comment

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/aws-c-common?pullRequest=1) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/aws-c-common?pullRequest=1) before we can accept your contribution.<br/>**0** out of **5** committers have signed the CLA.<br/><br/>:x: TingDaoK<br/>:x: graebm<br/>:x: TwistedTwigleg<br/>:x: bretambrose<br/>:x: DmitriyMusatkin<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/aws-c-common?pullRequest=1) it.</sub>

graebm committed:
fix MSVC warning (#912)

this was only appearing on MSVC Debug builds, with BUILD_TESTING=ON

TwistedTwigleg committed:
Make sure file paths are not empty strings to avoid segfault on Windows (#915)

Fixes segfaults that occur on Windows if you pass an empty string as a file path in the CRTs.

Commit log:
* Make sure file paths are not empty strings to avoid segfault on Windows
* Return NULL instead of raising an error in fopen and to_long_path
* Code review changes: Added helper functions and cleaned up code
* Renamed file to check if the string is empty rather than not empty
* Add back directory check I accidentally removed

TingDaoK committed:
reset the logger before memory trace (#917)

angelonakos committed:
Update CBMC starter kit to v2.5

angelonakos committed:
Remove starter-kit and Litani submodules

angelonakos committed:
Update README to include a section on CBMC

TingDaoK committed:
hide json module init and cleanup (#920)

gearnode committed:
Fix FreeBSD build (#921)

bretambrose committed:
Remove 'NOT' from a conditional (#922)

`NOT` is not a bash keyword.
The negation is done via '!'

Co-authored-by: Marco Magdy <marcomagdy@users.noreply.github.com>

TingDaoK committed:
Make libnuma linux only (#923)

TingDaoK committed:
Fall back for load target (#924)

- To fix load target based on the `BUILD_SHARED_LIBS`, which will lead to a problem when the lib was built with `BUILD_SHARED_LIBS` on, and the app built with `BUILD_SHARED_LIBS` off and the app will fail to load the path.
- Double fallback for now, we will introduce new cmake config to let user decide to enforce the shared/static lib to load.

graebm committed:
Hide symbols from libcrypto.a (#925)

**ISSUE:**
We were unable to take the latest version of AWS-LC (source of libcrypto.a).

AWS-LC changed the default visibility of symbols in static builds, from "hidden" to "visible" (see https://github.com/awslabs/aws-lc/pull/466)

This resulted in the symbols being visible in whatever shared library or executable ultimately used libcrypto.a. If an application ALSO ended up using libcrypto.so from the system's OpenSSL installation, we'd get crazy crashes due to SOME libcrypto function calls routing through libcrypto.a and SOME function calls routing through libcrypto.so.

**SOLUTION:**
Adapt the CMake function we use for configuring all C libraries in the Common Runtime ecosystem so that it always hides symbols from libcrypto.a.

DmitriyMusatkin committed:
Json iterators (#926)

add json functions to iterate over object fields/elements of array

bretambrose committed:
Safely handle a particular use case for linked hash tables where key equality is not reference equality (#928)

* Safely handle a particular use case for linked hash tables where key equality is not reference equality
* Add tests to cover the previously unusable cases

DmitriyMusatkin committed:
Json iterate signature (#929)

* change sig of json iterate callback to give more control over iteration

TwistedTwigleg committed:
Add AppVerifier (#930)

Adds a new AppVerifier CI task that will run alongside CTest to make sure Win32 API is being used correctly.

bretambrose committed:
Block logging updates (#931)

* Additional logging API/macro for better block logging

bretambrose committed:
Verify backtrace is defined too, not just header presence (#935)

* Verify backtrace is defined too, not just header presence
* clang tidy issues that only just now popped up

DmitriyMusatkin committed:
Uri migration (#936)

* migrate uri helpers from aws-c-io

DmitriyMusatkin committed:
json compare (#938)

* json compare

DmitriyMusatkin committed:
fix parsing of ipv6 literals in uris (#939)

* fix parsing of ipv6 literals in uris

DmitriyMusatkin committed:
Json duplicate (#940)

* add json duplicate api

DmitriyMusatkin committed:
parse uri path as is (#941)

DmitriyMusatkin committed:
make aws_json_value_destroy noop on null values (#942)

* make aws_json_value_destroy noop on null values

DmitriyMusatkin committed:
clarify aws_byte_cursor_next_split comments and add tests (#943)

* clarify aws_byte_cursor_next_split comments and add tests

graebm committed:
sprintf() -> snprintf() (#946)

Issue: Today my Mac started issuing warnings about sprintf().
Changes: Use snprintf() instead.

graebm committed:
Set thread names, for debugging (#937)

Also, rename a few choice functions where people often think our code is deadlocked, but it's not.

Using CamelCase names like "AwsHostResolver" because 15 characters is the max for Linux, so we can't do "aws_host_resolver"
